File: algorithms/cpp/superEggDrop/superEggDrop.cpp
// Source: https://leetcode.com/problems/super-egg-drop/
// Author: Miao Zhang
// Date:   2021-03-22

class Solution {
public:
    int superEggDrop(int k, int n) {
        m_ = vector<vector<int>>(k + 1, vector<int>(n + 1, INT_MIN));
        return dp(k, n);
    }

private:
    vector<vector<int>> m_;

    int dp(int k, int n) {
        if (k <= 0) return 0;
        if (k == 1) return n;
        if (n <= 1) return n;
        if (m_[k][n] != INT_MIN) return m_[k][n];
        int l = 1;
        int r = n + 1;
        while (l < r) {
            int m = l + (r - l) / 2;
            int broken = dp(k - 1, m - 1);
            int unbroken = dp(k, n - m);
            if (broken >= unbroken) {
                r = m;
            } else {
                l = m + 1;
            }
        }
        return m_[k][n] = 1 + dp(k - 1, l - 1);
    }
};

File: algorithms/python/superEggDrop/superEggDrop.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/super-egg-drop/
# Author: Miao Zhang
# Date:   2021-03-22

class Solution:
    def superEggDrop(self, k: int, n: int) -> int:
        self.m = [[float('-inf') for _ in range(n + 1)] for _ in range(k + 1)]
        return self.dp(k, n)

    def dp(self, k: int, n: int) -> int:
        if k <= 0: return 0
        if k <= 1: return n
        if n <= 1: return n
        if self.m[k][n] != float('-inf'):
            return self.m[k][n]
        l = 1
        r = n + 1
        while l < r:
            m = l + (r - l) // 2
            broken = self.dp(k - 1, m - 1)
            unbroken = self.dp(k, n - m)
            if broken >= unbroken:
                r = m
            else:
                l = m + 1
        self.m[k][n] = 1 + self.dp(k - 1, l - 1)
        return self.m[k][n]
